### ISSUE TITLE
RdfXml Parser question

### ISSUE THREAD

**Issue by philbartoncabi** [dotnetrdf/dotnetrdf#59]
Hello,

I am using RDFDotNet to parser RdfXml. I noticed an issue with the following XML being converted into a graph object:

&lt;rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"&gt;
    &lt;Concept rdf:about="http://id.cabi.org/cabt/61160" xmlns="http://www.w3.org/2004/02/skos/core#"&gt;
        &lt;narrower rdf:resource="http://id.cabi.org/cabt/107288"/&gt;
        &lt;narrower rdf:resource="http://id.cabi.org/cabt/107805"/&gt;
        &lt;narrower rdf:resource="http://id.cabi.org/cabt/94360"/&gt;
        &lt;narrower rdf:resource="http://id.cabi.org/cabt/97284"/&gt;
        &lt;rdf:type rdf:resource="http://www.w3.org/2004/02/skos/core#Concept"/&gt;
        &lt;created rdf:datatype="http://www.w3.org/2001/XMLSchema#string" xmlns="http://purl.org/dc/terms/"&gt;2004-08-03&lt;/created&gt;
        &lt;dateAccepted rdf:datatype="http://www.w3.org/2001/XMLSchema#string" xmlns="http://purl.org/dc/terms/"&gt;2008-08-19&lt;/dateAccepted&gt;
        &lt;modified rdf:datatype="http://www.w3.org/2001/XMLSchema#string" xmlns="http://purl.org/dc/terms/"&gt;2014-10-29&lt;/modified&gt;
        &lt;subjectCategory rdf:datatype="http://www.w3.org/2001/XMLSchema#string" xmlns="http://id.cabi.org/cabrelations/"&gt;Organism Names&lt;/subjectCategory&gt;
        &lt;GUID rdf:datatype="http://www.w3.org/2001/XMLSchema#string"&gt;{78A29560-3DD6-44E8-9F24-1D592260A42B}&lt;/GUID&gt;
        &lt;altLabel xml:lang="en-gb"&gt;Insecta&lt;/altLabel&gt;
        &lt;altLabel xml:lang="en-gb"&gt;insect&lt;/altLabel&gt;
        &lt;prefLabel xml:lang="en-gb"&gt;insects&lt;/prefLabel&gt;
    &lt;/Concept&gt;
&lt;/rdf:RDF&gt;

When the parser converts it into a graph, the namespace URI for the altLabel, prefLabel and GUID elements is set to http://id.cabi.org/cabrelations/, the last namespace declared on an element (subjectCategory) within the XML.

However, I believe that these should be within the http://www.w3.org/2004/02/skos/core# namespace.  Indeed, if I try this XML at the W3C validator, it puts these in this namespace.

Is this a bug or am I doing something wrong in the parsing of the XML.  At the moment my code is simply:

RdfXmlParser parser = new RdfXmlParser();

using (TextReader textReader = new StringReader(rdfXml))
{
   parser.Load(graph, textReader); 
}

Any help you could provide would be appreciated.

Cheers

Philip Barton.

**Comment by kal** [comment]
Hi Philip,

That does sound like a bug in the parser. The change to the default namespace should go out of scope at the `</subjectCategory>` close tag. 

I've replicated this in our test suite so I'll take a look at getting a fix out for it. The problem I've found is that namespaces are only ever added to the parser context as child elements are processed and they are not removed when the child element goes out of scope. 

As a workaround for now, is it possible for you to move all your namespace declarations up to the root element ? That would avoid getting into the situation of having namespaces going out of scope and being "overwritten" by namespaces defined on a parent element.

Cheers

Kal
